lib/Service/MessageOperationService.php

```diff
@@ -0,0 +1,153 @@
+
+class MessageOperationService {
+
+	public function __construct(
+		protected IMAPClientFactory $clientFactory,
+		protected MailAccountMapper $accountMapper,
+		protected MailboxMapper $mailboxMapper,
+		protected MessageMapper $messageMapper,
+		protected MailManager $mailManager,
+		protected ImapMessageMapper $imapMessageMapper,
+	) {
+	}
+
+	/**
+	 * convert message collection to grouped collections by mailbox id
+	 *
+	 * [[mailbox_id, uid, id]] to [mailbox_id => [[id, uid]]]
+	 *
+	 * @param array<array{0:int,1:int,2:int}> $collection
+	 *
+	 * @return array<int,array<array{0:int,1:int}>>
+	 */
+	protected function groupByMailbox(array $collection): array {
+		return array_reduce($collection, function ($carry, $pair) {
+			if (!isset($carry[$pair['mailbox_id']])) {
+				$carry[$pair['mailbox_id']] = [];
+			}
+			$carry[$pair['mailbox_id']][] = ['id' => $pair['id'], 'uid' => $pair['uid']];
+			return $carry;
+		}, []);
+	}
+
+	/**
+	 * convert mailbox collection to grouped collections by account id
+	 *
+	 * [mailbox] to [account_id => [mailbox]]
+	 *
+	 * @param array<\OCA\Mail\Db\MailBox> $collection
+	 *
+	 * @return array<int,array<\OCA\Mail\Db\MailBox>>
+	 */
+	protected function groupByAccount(array $collection) {
+		return array_reduce($collection, function ($carry, $entry) {
+			if (!isset($carry[$entry->getAccountId()])) {
+				$carry[$entry->getAccountId()] = [];
+			}
+			$carry[$entry->getAccountId()][] = $entry;
+			return $carry;
+		}, []);
+	}
+
+	/**
+	 * generates operation status responses for each message
+	 *
+	 * @param array<int,bool> &$results
+	 * @param bool $value
+	 * @param array<\OCA\Mail\Db\MailBox> $mailboxes
+	 * @param array<int,array<array{0:int,1:int}>> $messages
+	 */
+	protected function generateResult(array &$results, bool $value, array $mailboxes, array $messages) {
+		foreach ($mailboxes as $mailbox) {
+			foreach ($messages[$mailbox->getId()] as $message) {
+				$results[$message['id']] = $value;
+			}
+		}
+	}
+
+	/**
+	 * Set/Unset system flags or keywords
+	 *
+	 * @param string $userId system user id
+	 * @param array<int> $identifiers message ids
+	 * @param array<string,bool> $flags message flags
+	 *
+	 * @return array<int,bool> operation results
+	 */
+	public function changeFlags(string $userId, array $identifiers, array $flags): array {
+
+		// retrieve message meta data [uid, mailbox_id] for all messages and group by mailbox id
+		$messages = $this->groupByMailbox($this->messageMapper->findMailboxAndUid($identifiers));
+		// retrieve all mailboxes and group by account
+		$mailboxes = $this->groupByAccount($this->mailboxMapper->findByIds(array_keys($messages)));
+		// retrieve all accounts
+		$accounts = $this->accountMapper->findByIds(array_keys($mailboxes));
+		// process every account
+		$results = [];
+		foreach ($accounts as $account) {
+			$account = new Account($account);
+			// determine if account belongs to the user and skip if not
+			if ($account->getUserId() != $userId) {
```

Comment:
Good to check, but please move this down into the db when looking up accounts. Make it a WHERE id IN (xxx) AND user_id = 'sebastian'. That avoids loaded any foreign accounts.